File: spec/models/pattern_spec.rb
# frozen_string_literal: true

require 'rails_helper'
#
#  Pattern
#
#  id: int, primary key
#  account_id: int, foreign key
#  match_text: string
#  category_id: int, foreign key
#  subcategory_id: int foreign key
#
RSpec.describe Pattern do
  it 'has a valid factory' do
    p = FactoryBot.create(:pattern)
    expect(p).to be_valid
    expect(p).to be_a(described_class)
  end

  describe 'validations' do
    it 'is invalid without an account' do
      expect(FactoryBot.build(:pattern, account: nil)).not_to be_valid
    end

    it 'is invalid without a category' do
      expect(FactoryBot.build(:pattern, category: nil)).not_to be_valid
    end

    it 'is invalid without match text' do
      expect(FactoryBot.build(:pattern, match_text: nil)).not_to be_valid
    end

    it 'is valid without notes' do
      expect(FactoryBot.build(:pattern, notes: nil)).to be_valid
    end
  end

  describe 'relationships' do
    it 'belongs to account' do
      a = FactoryBot.create(:account)
      expect(FactoryBot.create(:pattern, account: a).account).to eq(a)
    end

    it 'belongs to category' do
      c = FactoryBot.create(:category)
      expect(FactoryBot.create(:pattern, category: c).category).to eq(c)
    end

    it 'belongs to subcategory' do
      s = FactoryBot.create(:subcategory)
      expect(FactoryBot.create(:pattern, subcategory: s).subcategory).to eq(s)
    end
  end
end

File: spec/javascripts/models/pattern_spec.js
describe('Pattern', function(){
  var pattern;
  beforeEach(function(){
    pattern = new MyMoney.Models.Pattern({id: 10, account_id: 14});
  });

  it('has a name', function(){
    expect(pattern.name).toEqual('pattern');
  })

  it('has a URL', function(){
    expect(pattern.url()).toEqual('accounts/14/patterns/10');
  });

  describe('validation', function(){
    it ('requires a match text', function(){
      expect(pattern.preValidate('match_text', '   ')).toEqual('Match text is required');
      expect(pattern.preValidate('match_text', 'anything')).toEqual('');
    });

    it ('requires a category', function(){
      expect(pattern.preValidate('category_id', '   ')).toEqual('Category is required');
      expect(pattern.preValidate('category_id', 60)).toEqual('');
    });

    it ('requires an account', function(){
      expect(pattern.preValidate('account_id', null)).toEqual('Account is required');
      expect(pattern.preValidate('account_id', 2)).toEqual('');
    });
  });
});